shell/platform/windows/game_pad_winuwp.cc

```diff
@@ -0,0 +1,228 @@
+  left_stick_callback_ = leftstick;
+  right_stick_callback_ = rightstick;
+
+  left_trigger_callback_ = lefttrigger;
+  right_trigger_callback_ = righttrigger;
+
+  button_pressed_callback_ = pressedcb;
+  button_released_callback_ = releasedcb;
+
+  arrival_departure_callback_ = changedcb;
+}
+
+void GamepadWinUWP::Initialize() {
+  RefreshCachedGamepads();
+
+  winrt::Windows::Gaming::Input::Gamepad::GamepadAdded(
+      {this, &GamepadWinUWP::OnGamepadAdded});
+  winrt::Windows::Gaming::Input::Gamepad::GamepadRemoved(
+      {this, &GamepadWinUWP::OnGamepadRemoved});
+
+  const std::lock_guard<std::mutex> lock(gamepad_mutex_);
+  current_game_pad_ = GetLastGamepad();
+}
+
+bool GamepadWinUWP::HasController() {
+  const std::lock_guard<std::mutex> lock(gamepad_mutex_);
+  return current_game_pad_ != nullptr;
+}
+
+void GamepadWinUWP::OnGamepadAdded(
+    winrt::Windows::Foundation::IInspectable const& th,
+    winrt::Windows::Gaming::Input::Gamepad const& args) {
+  enumerated_game_pads_.push_back(args);
+
+  {
+    const std::lock_guard<std::mutex> lock(gamepad_mutex_);
+    auto most_recent = GetLastGamepad();
+    if (current_game_pad_ != most_recent) {
+      current_game_pad_ = most_recent;
+    }
+  }
+
+  if (this->arrival_departure_callback_ != nullptr) {
+    arrival_departure_callback_();
+  }
+}
+
+void GamepadWinUWP::OnGamepadRemoved(
+    winrt::Windows::Foundation::IInspectable const&,
+    winrt::Windows::Gaming::Input::Gamepad const& /*args*/) {
+  RefreshCachedGamepads();
+}
+
+void GamepadWinUWP::RefreshCachedGamepads() {
+  enumerated_game_pads_.clear();
+  auto gamepads = winrt::Windows::Gaming::Input::Gamepad::Gamepads();
+  for (auto gamepad : gamepads) {
+    enumerated_game_pads_.push_back(gamepad);
+  }
+}
+
+const winrt::Windows::Gaming::Input::Gamepad* GamepadWinUWP::GetLastGamepad() {
+  winrt::Windows::Gaming::Input::Gamepad* gamepad = nullptr;
+
+  if (!enumerated_game_pads_.empty()) {
+    gamepad = &enumerated_game_pads_.back();
+  }
+
+  return gamepad;
+}
+
+bool isValid(double value) {
+  if (value > 0.1 || value < -0.1) {
+    return true;
+  }
+  return false;
+}
+
+void GamepadWinUWP::Process() {
+  {
+    const std::lock_guard<std::mutex> lock(gamepad_mutex_);
+
+    if (current_game_pad_ == nullptr) {
+      return;
+    }
+
+    last_reading_ = current_game_pad_->GetCurrentReading();
+  }
+
+  int exitComboPressed = 0;
+
+  namespace gi = winrt::Windows::Gaming::Input;
+
+  GamepadButtonPressedInternal(last_reading_.Buttons);
+
+  // work out which values have changed since the last reading.
+  if (last_reading_.LeftThumbstickX != left_stick_x_value_) {
+    left_stick_x_value_ = last_reading_.LeftThumbstickX;
+  }
+
+  if (last_reading_.LeftThumbstickY != left_stick_y_value_) {
+    left_stick_y_value_ = last_reading_.LeftThumbstickY;
+  }
+
+  if (isValid(left_stick_x_value_) || isValid(left_stick_y_value_)) {
+    RaiseLeftStickMoved(left_stick_x_value_, left_stick_y_value_);
+  }
```

Comment:
Here and for the right stick: this looks like `RaiseLeftStickMoved` will be called so long as the values are between -0.1 and 0.1, even if they haven't changed. Is that expected?

The trigger code looks more like what I'd expect.

Reply:
Yes because you are holding the stick at a constant angle, hence the value is constant, but you want the cursor to keep moving so you need to keep sending an update at polling frequency for it to keep moving.  Make sense?